Scala.js: Concatenating a Char and String causes assertion failure

## Compiler version

3.0.1

## Minimized code

```Scala
object FooBar {
  def baz(mySeq: Seq[String]) = mySeq.map(':' + _)
}
```

## Output

```
java.lang.AssertionError: assertion failed
    scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
    dotty.tools.backend.sjs.JSCodeGen.genStringConcat(JSCodeGen.scala:2487)
    dotty.tools.backend.sjs.JSCodeGen.genPrimitiveOp(JSCodeGen.scala:2098)
    dotty.tools.backend.sjs.JSCodeGen.genApply(JSCodeGen.scala:1712)
    dotty.tools.backend.sjs.JSCodeGen.genStatOrExpr(JSCodeGen.scala:1390)
    dotty.tools.backend.sjs.JSCodeGen.genExpr(JSCodeGen.scala:1263)
    dotty.tools.backend.sjs.JSCodeGen.genBody$1(JSCodeGen.scala:1146)
    dotty.tools.backend.sjs.JSCodeGen.genMethodDef(JSCodeGen.scala:1151)
    dotty.tools.backend.sjs.JSCodeGen.genMethodWithCurrentLocalNameScope$$anonfun$1(JSCodeGen.scala:1118)
    dotty.tools.backend.sjs.ScopedVar$.withScopedVars(ScopedVar.scala:35)
    dotty.tools.backend.sjs.JSCodeGen.genMethodWithCurrentLocalNameScope(JSCodeGen.scala:1124)
    dotty.tools.backend.sjs.JSCodeGen.genMethod$$anonfun$1(JSCodeGen.scala:1022)
    dotty.tools.backend.sjs.ScopedVar$.withScopedVars(ScopedVar.scala:35)
    dotty.tools.backend.sjs.JSCodeGen.genMethod(JSCodeGen.scala:1023)
    dotty.tools.backend.sjs.JSCodeGen.genScalaClass$$anonfun$2(JSCodeGen.scala:341)
    scala.collection.immutable.List.foreach(List.scala:333)
    dotty.tools.backend.sjs.JSCodeGen.genScalaClass(JSCodeGen.scala:346)
    dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit$$anonfun$6$$anonfun$1(JSCodeGen.scala:217)
    dotty.tools.backend.sjs.ScopedVar$.withScopedVars(ScopedVar.scala:35)
    dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit$$anonfun$2(JSCodeGen.scala:221)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:333)
    dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit(JSCodeGen.scala:223)
    dotty.tools.backend.sjs.JSCodeGen.run(JSCodeGen.scala:152)
    dotty.tools.backend.sjs.GenSJSIR.run(GenSJSIR.scala:15)
    dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
    scala.collection.immutable.List.map(List.scala:246)
    dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
    dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
    dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
    dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
    dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
    dotty.tools.dotc.Run.compileUnits(Run.scala:231)
    dotty.tools.dotc.Run.compileSources(Run.scala:166)
    dotty.tools.dotc.Run.compile(Run.scala:150)
    dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
    dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
    dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
    sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
    sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
    sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
    sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
    sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
    sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
    sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
    sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:457)
    sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
    sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
    sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
    sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
    sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:412)
    sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
    sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:399)
    sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
    sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
    sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
    sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
    sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
    mill.scalalib.worker.ZincWorkerImpl.compileInternal(ZincWorkerImpl.scala:569)
    mill.scalalib.worker.ZincWorkerImpl.$anonfun$compileMixed0$1(ZincWorkerImpl.scala:357)
    mill.api.FixSizedCache.withCachedValue(FixSizedCache.scala:66)
    mill.scalalib.worker.ZincWorkerImpl.withCompilers(ZincWorkerImpl.scala:452)
    mill.scalalib.worker.ZincWorkerImpl.compileMixed0(ZincWorkerImpl.scala:356)
    mill.scalalib.worker.ZincWorkerImpl.compileMixed(ZincWorkerImpl.scala:324)
    mill.scalalib.ScalaModule.$anonfun$compile$2(ScalaModule.scala:170)
    mill.define.ApplyerGenerated.$anonfun$zipMap$9(ApplicativeGenerated.scala:21)
    mill.define.Task$MappedDest.evaluate(Task.scala:374)

```

## Expectation

It should not crash with an assertion error. Worked around it by using only Strings: `":" + _`

--- a/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
+++ b/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
@@ -2790,27 +2790,7 @@ class JSCodeGen()(using genCtx: Context) {
       args: List[Tree]): js.Tree = {
     implicit val pos = tree.span
 
-    val arg = args.head
-
-    /* Primitive number types such as scala.Int have a
-     *   def +(s: String): String
-     * method, which is why we have to box the lhs sometimes.
-     * Otherwise, both lhs and rhs are already reference types (Any or String)
-     * so boxing is not necessary (in particular, rhs is never a primitive).
-     */
-    assert(!isPrimitiveValueType(receiver.tpe) || arg.tpe.isRef(defn.StringClass))
-    assert(!isPrimitiveValueType(arg.tpe))
-
-    val genLhs = {
-      val genLhs0 = genExpr(receiver)
-      // Box the receiver if it is a primitive value
-      if (!isPrimitiveValueType(receiver.tpe)) genLhs0
-      else makePrimitiveBox(genLhs0, receiver.tpe)
-    }
-
-    val genRhs = genExpr(arg)
-
-    js.BinaryOp(js.BinaryOp.String_+, genLhs, genRhs)
+    js.BinaryOp(js.BinaryOp.String_+, genExpr(receiver), genExpr(args.head))
   }
 
   /** Gen JS code for a call to Any.## */

--- a/tests/sjs-junit/test/org/scalajs/testsuite/compiler/RegressionTestScala3.scala
+++ b/tests/sjs-junit/test/org/scalajs/testsuite/compiler/RegressionTestScala3.scala
@@ -44,6 +44,14 @@ class RegressionTestScala3 {
     assertEquals(1, X_Issue13221.I.i)
     assertEquals(1, X_Issue13221.blah)
   }
+
+  @Test def primitivePlusStringThatIsATermRefIssue13518(): Unit = {
+    def charPlusString(x: String): String = 'a' + x
+    assertEquals("abc", charPlusString("bc"))
+
+    def intPlusString(x: String): String = 5 + x
+    assertEquals("5bc", intPlusString("bc"))
+  }
 }
 
 object RegressionTestScala3 {

